src/platform.ts

```diff
@@ -54,197 +59,591 @@ export function getPlatformDetails(): IPlatformDetails {
+
+    // Windows PowerShell installations
+    private readonly winPSExeVal: Lazy<IPossiblePowerShellExe>;
+    private readonly alternateBitnessWinPSExeVal: Lazy<IPossiblePowerShellExe>;
+
+    // Additional configured PowerShells
+    private readonly additionalPSExeSettings: Iterable<Settings.IPowerShellAdditionalExePathSettings>;
+
+    /**
+     * Create a new PowerShellFinder object to discover PowerShell installations.
+     * @param platformDetails Information about the machine we are running on.
+     * @param additionalPowerShellExes Additional PowerShell installations as configured in the settings.
+     */
+    constructor(
+        platformDetails?: IPlatformDetails,
+        additionalPowerShellExes?: Iterable<Settings.IPowerShellAdditionalExePathSettings>) {
+
+        this.platformDetails = platformDetails || getPlatformDetails();
+        this.additionalPSExeSettings = additionalPowerShellExes || [];
+
+        this.pwshWindowsInstallationsVal = new Lazy(() => this.findPSCoreWindowsInstallations());
+        this.pwshAlternateBitnessWindowsInstallationsVal = new Lazy(
+            () => this.findPSCoreWindowsInstallations({ findNonNativeBitness: true }));
+
+        this.stablePwshExeVal                  = new Lazy(() => this.findPSCoreStable());
+        this.stableAlternateBitnessPwshExeVal  = new Lazy(() => this.findPSCoreAlternateBitnessStable());
+        this.previewAlternateBitnessPwshExeVal = new Lazy(() => this.findPSCoreAlternateBitnessPreview());
+        this.previewPwshExeVal                 = new Lazy(() => this.findPSCorePreview());
+        this.dotnetGlobalToolExeVal            = new Lazy(() => this.findPSCoreDotnetGlobalTool());
+        this.msixExeVal                        = new Lazy(() => this.findPSCoreMsix());
+        this.stableSnapExeVal                  = new Lazy(() => this.findPSCoreStableSnap());
+        this.previewSnapExeVal                 = new Lazy(() => this.findPSCorePreviewSnap());
+        this.winPSExeVal                       = new Lazy(() => this.findWinPS());
+        this.alternateBitnessWinPSExeVal       = new Lazy(() => this.findWinPS({ findNonNativeBitness: true }));
+    }
+
+    /**
+     * The stable PowerShell 6+ installation.
+     * May be null if the installation directory is not present.
+     */
+    private get pwshStable(): IPossiblePowerShellExe {
+        return this.stablePwshExeVal.value;
+    }
+
+    /**
+     * The preview PowerShell 6+ installation.
+     * May be null if the installation directory is not present.
+     */
+    private get pwshPreview(): IPossiblePowerShellExe {
+        return this.previewPwshExeVal.value;
+    }
+
+    /**
+     * The stable non-process-native bitness PowerShell 6+ installation.
+     * This means 32-bit in a 64-bit process, and vice versa.
+     * May be null if the installation directory is not present
+     */
+    private get pwshAlternateBitnessStable(): IPossiblePowerShellExe {
```

Comment:
They provide a non-lazy interface to a lazy value. Can change to just using the lazy field if you prefer

Reply:
I guess the whole concept of the lazy fields just throws me off when secretly, they're just calling `find*()`.

Adding the lazy fields added like 200 lines of code in this file and I'm wondering if that's really cleaner than calling the `find*()` methods directly in `enumerateDefaultPowerShellInstallations`.

Unless I'm missing something?

Reply:
I did it to make the final enumerate method easy to read so it's obvious what our intended discovery order is. But I'll take it out I suppose.

Reply:
Compared to the `find*` methods though, the lazy values allow us to cache things, while only doing the work we need to do at startup.

I originally had variables in the find methods but decided to factor it out since it was repetitive and ugly

Reply:
It's fine. Although that does add a chunk of code to the file, I suppose it doesn't add too much real complexity.